kbmd/kbmd.c

```diff
@@ -0,0 +1,470 @@
+		kbmd_dfatal(dfd, "could not set default system pool: %s",
+		    strerror(errno));
+	}
+
+	mutex_exit(&piv_lock);
+
+	kbmd_event_init(dfd);
+	kbmd_recover_init(dfd);
+	load_plugin();
+
+	(void) bunyan_trace(tlog, "Creating door server", BUNYAN_T_END);
+
+	errval = kbmd_door_setup(doorpath);
+	if (errval != 0)
+		kbmd_dfatal(dfd, "unable to create door");
+
+	errval = 0;
+	(void) write(dfd, &errval, sizeof (errval));
+	(void) close(dfd);
+
+	for (;;) {
+		if (sigsuspend(&set) == -1)
+			if (errno == EFAULT)
+				abort();
+		if (kbmd_quit == B_TRUE)
+			break;
+	}
+
+	(void) bunyan_info(tlog, "Exiting...", BUNYAN_T_END);
+
+	kbmd_event_fini();
+	return (0);
+}
+
+/*
+ * We borrow fmd's daemonization style. Basically, the parent waits for the
+ * child to successfully set up a door and recover all of the old configurations
+ * before we say that we're good to go.
+ */
+static int
+kbmd_daemonize(int dirfd)
+{
+	char path[PATH_MAX];
+	struct rlimit rlim;
+	sigset_t set, oset;
+	int estatus, pfds[2];
+	pid_t child;
+#if notyet
+	priv_set_t *pset;
+#endif
+
+	/*
+	 * Set a per-process core path to be inside of /var/run/kbmd. Make sure
+	 * that we aren't limited in our dump size.
+	 */
+	VERIFY3S(snprintf(path, sizeof (path),
+	    "%s/core.%s.%%p", KBMD_RUNDIR, getprogname()), >, 0);
+	(void) core_set_process_path(path, strlen(path) + 1, getpid());
+
+	rlim.rlim_cur = RLIM_INFINITY;
+	rlim.rlim_max = RLIM_INFINITY;
+	if (setrlimit(RLIMIT_CORE, &rlim) < 0)
+		warn("unable to set core file size to unlimited");
+
+	/*
+	 * Claim as many file descriptors as the system will let us.
+	 */
+	if (getrlimit(RLIMIT_NOFILE, &rlim) == 0) {
+		rlim.rlim_cur = rlim.rlim_max;
+		(void) setrlimit(RLIMIT_NOFILE, &rlim);
+	}
+
+	/*
+	 * chdir /var/run/varpd
+	 */
+	if (fchdir(dirfd) != 0)
+		err(EXIT_FAILURE, "failed to chdir to %s", KBMD_RUNDIR);
+
+
+	/*
+	 * At this point block all signals going in so we don't have the parent
+	 * mistakingly exit when the child is running, but never block SIGABRT.
+	 */
+	if (sigfillset(&set) != 0)
+		abort();
+	if (sigdelset(&set, SIGABRT) != 0)
+		abort();
+	if (sigprocmask(SIG_BLOCK, &set, &oset) != 0)
+		abort();
+
+	/*
+	 * Do the fork+setsid dance.
+	 */
+	if (pipe(pfds) != 0)
+		err(EXIT_FAILURE, "failed to create pipe for daemonizing");
+
+	if ((child = fork()) == -1)
+		err(EXIT_FAILURE, "failed to fork for daemonizing");
+
+	if (child != 0) {
+		/* We'll be exiting shortly, so allow for silent failure */
+		(void) close(pfds[1]);
+		if (read(pfds[0], &estatus, sizeof (estatus)) ==
+		    sizeof (estatus))
+			_exit(estatus);
+
+		if (waitpid(child, &estatus, 0) == child && WIFEXITED(estatus))
+			_exit(WEXITSTATUS(estatus));
+
+		_exit(EXIT_FAILURE);
+	}
+
+	/*
+	 * Drop privileges here.
+	 *
+	 * TODO: Determine what things are needed/can be dropped
+	 */
+	if (setgroups(0, NULL) != 0)
+		abort();
+	if (setgid(GID_KBMD) == -1 || seteuid(UID_KBMD) == -1)
+		abort();
+#if notyet
+	if ((pset = priv_allocset()) == NULL)
```

Comment:
It's gone -- with having to run plugins, which themselves run other stuff (curl, sysinfo, etc.) I couldn't get a narrower set of privileges where everything still worked unfortunately (sysinfo seemed to be the big culprit).. perhaps in a future change (certainly not this one) sysinfo could turn into a small daemon that gathers everything, and then just communicates via a door, so that 'callers' of sysinfo don't need full privileges.. and then it could make sense to revisit this.

Reply:
confirmed